openedx/core/djangoapps/content/learning_sequences/api/tests/test_outlines.py

```diff
@@ -1302,15 +1381,141 @@ def test_roundtrip(self):
+        ),
+        (
+            None,
+            set([2, 7]),
+            [CourseMode.VERIFIED, CourseMode.MASTERS],
+            {'student1': 'verified', 'student2': 'masters'},
+            {'student1': 1, 'student2': 1}
+        ),
+        (
+            None,
+            set([2]),
+            [CourseMode.VERIFIED, CourseMode.MASTERS],
+            {'student1': 'verified', 'student2': 'masters'},
+            {'student1': 1, 'student2': 0}
+        ),
+        (
+            None,
+            set([7]),
+            [CourseMode.VERIFIED, CourseMode.MASTERS],
+            {'student1': 'verified', 'student2': 'masters'},
+            {'student1': 0, 'student2': 1}
+        ),
+    )
+    @ddt.unpack
+    def test_enrollment_track_partition_on_section(
+        self,
+        section_visible_groups,
+        sequence_visible_groups,
+        course_modes,
+        learners_with_modes,
+        expected_values_dict
+    ):
+        section_user_partition_groups = None
+        sequence_user_partition_groups = None
+        if section_visible_groups:
+            section_user_partition_groups = {50: frozenset(section_visible_groups)}
+        if sequence_visible_groups:
+            sequence_user_partition_groups = {50: frozenset(sequence_visible_groups)}
+
+        self._setup_course_outline_with_sections(
+            [
+                CourseSectionData(
+                    usage_key=self.course_key.make_usage_key('chapter', '0'),
+                    title="Section 0",
+                    user_partition_groups=section_user_partition_groups,
+                    sequences=[
+                        CourseLearningSequenceData(
+                            usage_key=self.course_key.make_usage_key('subsection', '0'),
+                            title='Subsection 0',
+                            visibility=self.visibility,
+                            user_partition_groups=sequence_user_partition_groups,
+                        ),
+                    ]
+                )
+            ]
+        )
+
+        for course_mode in course_modes:
+            self._add_course_mode(
+                self.course_key,
+                mode_slug=course_mode,
+                mode_display_name=course_mode,
+            )
+
+        # Enroll students in the course
+        learners_to_verify = set()
+        for username, mode in learners_with_modes.items():
+            learners_to_verify.add(
+                self._create_and_enroll_learner(username, mode)
+            )
+
+        check_date = datetime(2021, 3, 27, tzinfo=timezone.utc)
+
+        # Get details
+        staff_details, _, beta_tester_details = self.get_details(check_date)
+
+        assert len(staff_details.outline.accessible_sequences) == 1
+        assert len(beta_tester_details.outline.accessible_sequences) == 0
+
+        for learner_to_verify in learners_to_verify:
+            learner_details = get_user_course_outline_details(self.course_key, learner_to_verify, check_date)
```

Comment:
Doing this kind of high level testing is okay (and is the pattern for this module), but I would advocate that you try to do a test that's very specific to your outline processor, in its own module. That would mean instantiating your processor directly, and checking something like:

```python
processor = EnrollmentTrackPartitionGroupsOutlineProcessor(course_key, user, at_time)
processor.load_data()
removed_usage_keys = processor.usage_keys_to_remove(full_course_outline)
# check that the removed keys line up to your expectations
```

In that way, you don't have to worry about the other processors interfering in your test results (like with dates, for example). I hope to eventually move all the processors over to that style, since the tests in this module get more and more complex as the number of processors grows, and that's not a good pattern to set.

Reply:
I see. I'll write a new test suite class for this purpose. I'll leave what I have here intact because it would be useful still. I'll let you know once that's done.

Reply:
I added one more set of test cases to the original test suite for only testing the `EnrollmentTrackPartitionGroupsOutlineProcessor`. The code is largely the same. But I feel it's important to write a new test set as you suggested. Please review again.